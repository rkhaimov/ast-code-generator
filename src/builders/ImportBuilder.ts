import { ImportDeclarationStructure } from 'ts-simple-ast';

import { ProjectManager } from '../managers/ProjectManager';

class _ImportBuilder {
  importRepositoryBase(): ImportDeclarationStructure {
    return {
      moduleSpecifier: `./${ProjectManager.FILE_NAMES.REPOSITORY_BASE}`,
      namedImports: [{ name: ProjectManager.FILE_NAMES.REPOSITORY_BASE }],
    };
  }
}

export const ImportBuilder = new _ImportBuilder();
