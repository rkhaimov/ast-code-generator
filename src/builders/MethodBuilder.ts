import { isEmpty, last, keys, first, get } from 'lodash';
import { MethodDeclarationStructure, ParameterDeclarationStructure, SourceFile } from 'ts-simple-ast';
import escodegen from 'escodegen';

import { GetImplementationBuilder } from './implementation/methods/GetImplementationBuilder';
import { PostImplementationBuilder } from './implementation/methods/PostImplementationBuilder';
import { PutImplementationBuilder } from './implementation/methods/PutImplementationBuilder';

import { BlockStatementBody, IIdentifier } from '../definitions/ast/common';
import { ISwaggerMethod, ISwaggerMethodParam, ISwaggerOperations } from '../definitions/swagger';

interface IMethodBuilder {
  buildMethod(api: string, operations: ISwaggerOperations, definitions: SourceFile): MethodDeclarationStructure;
}

class _MethodBuilder implements IMethodBuilder {
  buildMethod(api: string, operations: ISwaggerOperations, definitions: SourceFile): MethodDeclarationStructure {
    const methodName = last<string>(api.split('/'));
    const methods: string[] = keys(operations);
    const method: ISwaggerMethod = get(operations, first(methods)!);

    return {
      name: methodName!,
      parameters: this.buildFunctionParams(method.parameters),
      bodyText: this.buildBodyText(api, operations),
    };
  }

  private buildBodyText(api: string, operations: ISwaggerOperations) {
    return escodegen.generate(this.buildFunctionBody(api, operations));
  }

  private buildFunctionBody(api: string, operations: ISwaggerOperations): BlockStatementBody {
    if ('get' in operations) {
      return GetImplementationBuilder.buildImplementation(api, operations.get);
    }

    if ('post' in operations) {
      return PostImplementationBuilder.buildImplementation(api, operations.post);
    }

    if ('put' in operations) {
      return PutImplementationBuilder.buildImplementation(api, operations.put);
    }

    if ('delete' in operations) {
      return PutImplementationBuilder.buildImplementation(api, operations.delete);
    }

    throw new Error(`Unknown operation(s) was provided ${keys(operations)}`);
  }

  private buildFunctionParams(swaggerParams: ISwaggerMethodParam[]): ParameterDeclarationStructure[] {
    if (isEmpty(swaggerParams)) {
      return [];
    }

    const param = {
      name: 'payload',
      type: 'Identifier',
    };

    return [param as IIdentifier];
  }
}

export const MethodBuilder: IMethodBuilder = new _MethodBuilder();
