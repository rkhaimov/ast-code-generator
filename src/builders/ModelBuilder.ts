/* tslint:disable:max-line-length */
import { reduce, last } from 'lodash';
import { InterfaceDeclarationStructure, PropertySignatureStructure } from 'ts-simple-ast';

import { ISwagger, SwaggerDefinitionPropertyTypes } from '../definitions/swagger';

class _ModelBuilder {
  buildModel(model: string, definition: ISwagger['definitions'][string]): InterfaceDeclarationStructure {
    const properties: InterfaceDeclarationStructure['properties'] = reduce(definition.properties, (acc, value: SwaggerDefinitionPropertyTypes, name) => {
      const property: PropertySignatureStructure = {
        name,
        type: this.toTsType(value),
      };

      return (acc as InterfaceDeclarationStructure['properties'])!.concat(property);
    }, [] as InterfaceDeclarationStructure['properties']);

    return {
      name: `I${model}`,
      properties,
      isExported: true,
    };
  }

  private toTsType(property: SwaggerDefinitionPropertyTypes): string {
    switch (property.type) {
      case 'integer': {
        return 'number';
      }
      case 'string': {
        if (property.enum) {
          return property.enum.map((value) => `'${value}'`).join(' | ');
        }

        return 'string';
      }
      case 'array': {
        if (!('$ref' in property.items)) {
          throw new Error('Unrecognizable array swagger structure');
        }

        const name = this.getRefDefinitionName(property.items.$ref);

        return `${name}[]`;
      }
      default: {
        return property.type;
      }
    }
  }

  private getRefDefinitionName(ref: string): string {
    return `I${last(ref.split('/'))}`;
  }
}

export const ModelBuilder = new _ModelBuilder();
