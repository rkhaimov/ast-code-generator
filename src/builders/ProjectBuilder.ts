export class ProjectBuilder {
  static FILE_NAMES = {
    DEFINITIONS: 'definitions',
    REPOSITORIES: 'repositories',
    REPOSITORY_BASE: 'RepositoryBase',
  };
}
