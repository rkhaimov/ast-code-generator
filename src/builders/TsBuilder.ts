import { ts } from 'ts-simple-ast';

class _TsBuilder {
  project = ts.createSourceFile('ts-build-source.ts', '', ts.ScriptTarget.ES2015, true, ts.ScriptKind.TS);
  printer = ts.createPrinter();

  print(node: ts.Node): string {
    return this.printer.printNode(ts.EmitHint.Unspecified, node, this.project);
  }
}

export const TsBuilder = new _TsBuilder();
