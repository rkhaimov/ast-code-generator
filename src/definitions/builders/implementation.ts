import { ISwaggerMethodParam } from '../swagger';

export type ArgumentsGroup = {
  [P in ISwaggerMethodParam['in']]: ISwaggerMethodParam;
};
