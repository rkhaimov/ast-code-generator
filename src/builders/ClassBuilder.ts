import { reduce } from 'lodash';
import { ClassDeclarationStructure, MethodDeclarationStructure, SourceFile } from 'ts-simple-ast';

import { MethodBuilder } from './MethodBuilder';

import { ISwagger } from '../definitions/swagger';

interface IClassBuilder {
  build(name: string, methods: ISwagger['paths'], definitions: SourceFile): ClassDeclarationStructure;
}

class _ClassBuilder implements IClassBuilder {
  build(name: string, methods: ISwagger['paths'], definitions: SourceFile): ClassDeclarationStructure {
    const body = reduce(methods, (acc, operation, api) => {
      const method = MethodBuilder.buildMethod(api, operation, definitions);

      return acc.concat(method);
    }, [] as MethodDeclarationStructure[]);

    return {
      name: `${name}Repository`,
      extends: 'RepositoryBase',
      methods: body,
      isExported: true,
    };
  }
}

export const ClassBuilder: IClassBuilder = new _ClassBuilder();
