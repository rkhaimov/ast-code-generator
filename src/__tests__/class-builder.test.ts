import pick from 'lodash/pick';
import swaggerMock from './__mocks__/swagger.json';

import { ClassBuilder } from '../builders/ClassBuilder';

describe('Class builder works well when', () => {
  const paths = pick(swaggerMock.paths, ['/api/account', '/api/account/signin']);
  const className = 'Account';
  const ast = ClassBuilder.build(className, paths);

  it('generates proper className', () => {
    expect(ast.id!.name).toEqual(className);
  });
});
