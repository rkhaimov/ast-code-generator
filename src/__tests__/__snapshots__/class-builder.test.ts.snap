// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Class builder works well when matches giving code style 1`] = `
"class AccountRepository extends RepositoryBase {
    account() {
        return this.get('/api/account');
    }

    signin(payload: Identifier) {
        return this.post('/api/account/signin', this.pick(payload, ['request']));
    }

    agreement(payload: Identifier) {
        return this.get(\`\${ '/api/agreement' }\${ this.toQuery(this.pick(payload, [
            'Initiator',
            'DeadlineFrom',
            'DeadlineTo',
            'Title',
            'CreatedFrom',
            'CreatedTo',
            'Status',
            'AssignedTo'
        ])) }\`);
    }

    details(payload: Identifier) {
        return this.get(this.fillPath('/api/agreement/{id}/details', payload));
    }

    delegate(payload: Identifier) {
        return this.post(this.fillPath('/api/outsourcing/{id}/delegate', payload), this.pick(payload, ['request']));
    }
}
"
`;
